pkg/pair/pair_test.go

```diff
@@ -0,0 +1,285 @@
+	"encoding/base64"
+	"encoding/csv"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"math/big"
+	"optable-pair-cli/pkg/keys"
+	"strings"
+	"testing"
+
+	"github.com/optable/match/pkg/pair"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPAIRIDReadWriter_HashEncrypt(t *testing.T) {
+	t.Parallel()
+	// arrange
+	lenEmails := 1001
+	ctx := context.Background()
+	salt := requireGenSalt(t)
+	key := requireGenKey(t)
+	emails := requireGenRandomHashedEmails(t, lenEmails)
+	expected := requireEncryptEmails(t, emails, salt, key)
+	r, w := bytes.NewBuffer(nil), bytes.NewBuffer(nil)
+
+	// set emails in csv format for PAIRIDReadWriter to read
+	requireWriteEmails(t, r, emails)
+
+	// act
+	rw, err := NewPAIRIDReadWriter(r, w)
+	require.NoError(t, err, "must create PAIRIDReadWriter")
+
+	err = rw.HashEncrypt(ctx, 1, salt, key)
+	require.NoError(t, err, "must hash and encrypt emails")
+
+	// assert
+	csvData := csv.NewReader(w)
+	hashEncryptedData, err := csvData.ReadAll()
+	require.NoError(t, err, "must read csv data")
+	require.Len(t, hashEncryptedData, len(expected), "must contain all emails")
+
+	for i, hashEncrypted := range hashEncryptedData {
+		require.Len(t, hashEncrypted, 1, "must contain one csv column")
+		require.Equal(t, expected[i], hashEncrypted[0], "encrypted email must match")
+	}
+}
+
+func TestPAIRIDReadWriter_ReEncrypt(t *testing.T) {
+	t.Parallel()
+	// arrange
+	lenEmails := 10000
+	ctx := context.Background()
+	salt := requireGenSalt(t)
+	key := requireGenKey(t)
+	emails := requireGenRandomHashedEmails(t, lenEmails)
+	encryptedEmails := requireEncryptEmails(t, emails, salt, key)
+	twiceEncryptedEmails := requireReEncryptEmails(t, encryptedEmails, salt, key)
+	r, w := bytes.NewBuffer(nil), bytes.NewBuffer(nil)
+
+	// set twice encrypted emails in csv format for PAIRIDReadWriter to read
+	requireWriteEmails(t, r, encryptedEmails)
+
+	// in this test we check encrypted emails are encrypted correctly and shuffled
+	expected := twiceEncryptedEmails
+
+	// act
+	rw, err := NewPAIRIDReadWriter(r, w)
+	require.NoError(t, err, "must create PAIRIDReadWriter")
+
+	err = rw.ReEncrypt(ctx, 1, salt, key)
+	require.NoError(t, err, "must re-encrypt emails")
+
+	// assert
+	csvData := csv.NewReader(w)
+	reEncryptedData, err := csvData.ReadAll()
+	require.NoError(t, err, "must read csv data")
+	require.Len(t, reEncryptedData, len(expected), "must contain all emails")
+
+	notShuffled := 0
+	for i, reEncrypted := range reEncryptedData {
+		require.Len(t, reEncrypted, 1, "must contain one csv column")
+
+		// check how many emails stay at the same place
+		if reEncrypted[0] == expected[i] {
+			notShuffled++
+		}
+
+		// must find the encrypted email in the expected slice
+		found := false
+		for _, e := range expected {
+			if e == reEncrypted[0] {
+				found = true
+				break
+			}
+		}
+		require.True(t, found, "re-encrypted email must match")
+	}
+
+	require.Less(t, float64(notShuffled), float64(lenEmails)*0.01, "must shuffle more then 99% of emails")
```

Comment:
```suggestion
	require.Less(t, float64(notShuffled), float64(lenEmails)*0.01, "must shuffle more than 99% of emails")
```

Reply:
Wait, are we actually doing a shuffle? @juanli16 I thought the Shuffle is only done in the publisher side?